cache/disk/disk.go

```diff
@@ -264,55 +281,181 @@ func (c *diskCache) Put(kind cache.EntryKind, hash string, expectedSize int64, r
 
 	// Rename to the final path
 	err = os.Rename(tmpFilePath, filePath)
-	// Only commit if renaming succeeded
 	if err == nil {
+		// Only commit if renaming succeeded.
 		// This flag is used by the defer() block above.
 		shouldCommit = true
 	}
 
 	return err
 }
 
-func (c *diskCache) Get(kind cache.EntryKind, hash string) (rdr io.ReadCloser, sizeBytes int64, err error) {
-	if !c.Contains(kind, hash) {
+// Return true if the item is in the cache and available.
+// Otherwise if it's ok to proxy, return a non-nil uncommitted *lruItem.
+// The caller must either set the size and commit the item, or remove it
+// from the LRU.
+func (c *DiskCache) availableOrTryProxy(key string) (bool, *lruItem) {
+	inProgress := false
+	var newItem *lruItem
+
+	c.mu.Lock()
+
+	existingItem, found := c.lru.Get(key)
+	if found {
+		if !existingItem.(*lruItem).committed {
+			inProgress = true
+		}
+	} else if c.proxy != nil {
+		newItem = &lruItem{
+			size:      -1, // Caller must fill this in later!
+			committed: false,
+		}
+		c.lru.Add(key, newItem)
+	}
+
+	c.mu.Unlock()
+
+	available := found && !inProgress
+
+	return available, newItem
+}
+
+// Get returns an io.ReadCloser with the content of the cache item stored under `hash`
+// and the number of bytes that can be read from it. If the item is not found, the
+// io.ReadCloser will be nil. If some error occurred when processing the request, then
+// it is returned.
+func (c *DiskCache) Get(kind cache.EntryKind, hash string) (io.ReadCloser, int64, error) {
+
+	var err error
+	key := cacheKey(kind, hash)
+
+	available, newItem := c.availableOrTryProxy(key)
+
+	if available {
+		blobPath := cacheFilePath(kind, c.dir, hash)
+		fileInfo, err := os.Stat(blobPath)
+		if err == nil {
+			r, err := os.Open(blobPath)
+			if err == nil {
+				cacheHits.Inc()
+				return r, fileInfo.Size(), nil
+			}
+		}
+
 		cacheMisses.Inc()
-		return nil, 0, nil
+		return nil, -1, nil
 	}
 
-	blobPath := cacheFilePath(kind, c.dir, hash)
+	cacheMisses.Inc()
+
+	if newItem == nil {
+		return nil, -1, nil
+	}
 
-	fileInfo, err := os.Stat(blobPath)
+	filePath := cacheFilePath(kind, c.dir, hash)
+	tmpFilePath := filePath + ".tmp"
+	shouldCommit := false
+	tmpFileCreated := false
+	foundSize := int64(-1)
+	var f *os.File
+
+	// We're allowed to try downloading this blob from the proxy.
+	// Before returning, we have to either commit the item and set
+	// its size, remove the item from the LRU.
+	defer func() {
+		c.mu.Lock()
+
+		if shouldCommit {
+			newItem.committed = true
+			newItem.size = foundSize
+		} else {
+			c.lru.Remove(key)
+		}
+
+		c.mu.Unlock()
+
+		if !shouldCommit && tmpFileCreated {
+			os.Remove(tmpFilePath) // No need to check the error.
+		}
+
+		f.Close() // No need to check the error.
+	}()
+
+	r, foundSize, err := c.proxy.Get(kind, hash)
+	if err != nil || r == nil {
+		return nil, -1, err
+	}
+	defer r.Close()
```

Comment:
Could this possibly leak a file descriptor if there is an err but r is not nil?

Consider

```
r, foundSize, err := ...
if r != nil {
  defer r.Close()
}
```

Reply:
Yes, good call. Fixed.